polkadot/xcm/pallet-xcm/src/lib.rs

```diff
@@ -1256,31 +1358,178 @@ impl<T: Config> Pallet<T> {
 		let value = (origin_location, assets.into_inner());
 		ensure!(T::XcmTeleportFilter::contains(&value), Error::<T>::Filtered);
 		let (origin_location, assets) = value;
+		for asset in assets.iter() {
+			let transfer_type =
+				<T::XcmExecutor as AssetTransferSupport>::determine_for(asset, &dest)
+					.map_err(Error::<T>::from)?;
+			ensure!(matches!(transfer_type, TransferType::Teleport), Error::<T>::Filtered);
+		}
+		let fees = assets.get(fee_asset_item as usize).ok_or(Error::<T>::Empty)?.clone();
+
+		Self::build_and_execute_xcm_transfer_type(
+			origin_location,
+			dest,
+			beneficiary,
+			assets,
+			TransferType::Teleport,
+			fees,
+			false,
+			weight_limit,
+		)
+	}
+
+	fn build_and_execute_xcm_transfer_type(
+		origin: impl Into<MultiLocation>,
+		dest: MultiLocation,
+		beneficiary: MultiLocation,
+		assets: Vec<MultiAsset>,
+		transfer_type: TransferType,
+		fees: MultiAsset,
+		jit_withdraw_fee_on_dest: bool,
+		weight_limit: WeightLimit,
+	) -> DispatchResult {
+		let mut message = match transfer_type {
+			TransferType::LocalReserve => Self::local_reserve_transfer_message(
+				dest,
+				beneficiary,
+				assets,
+				fees,
+				jit_withdraw_fee_on_dest,
+				weight_limit,
+			),
+			TransferType::DestinationReserve => Self::destination_reserve_transfer_message(
+				dest,
+				beneficiary,
+				assets,
+				fees,
+				jit_withdraw_fee_on_dest,
+				weight_limit,
+			),
+			TransferType::RemoteReserve(reserve) => Self::remote_reserve_transfer_message(
+				reserve,
+				dest,
+				beneficiary,
+				assets,
+				fees,
+				weight_limit,
+			),
+			TransferType::Teleport =>
+				Self::teleport_asset_message(dest, beneficiary, assets, fees, weight_limit),
+		}?;
+		let weight =
+			T::Weigher::weight(&mut message).map_err(|()| Error::<T>::UnweighableMessage)?;
+		let hash = message.using_encoded(sp_io::hashing::blake2_256);
+		let outcome = T::XcmExecutor::execute_xcm_in_credit(origin, message, hash, weight, weight);
+		Self::deposit_event(Event::Attempted { outcome });
+		Ok(())
+	}
+
+	fn local_reserve_transfer_message(
+		dest: MultiLocation,
+		beneficiary: MultiLocation,
+		assets: Vec<MultiAsset>,
+		mut fees: MultiAsset,
+		jit_withdraw_fee_on_dest: bool,
+		weight_limit: WeightLimit,
+	) -> Result<Xcm<<T as Config>::RuntimeCall>, Error<T>> {
+		let context = T::UniversalLocation::get();
+		fees.reanchor(&dest, context).map_err(|_| Error::<T>::CannotReanchor)?;
+		let max_assets = assets.len() as u32;
+		let xcm_on_dest = Xcm(vec![
+			SetFeesMode { jit_withdraw: jit_withdraw_fee_on_dest },
```

Comment:
Could you, please elaborate on the purpose of this `SetFeesMode` instruction (when `jit_withdraw_fee_on_dest` is `true`)? Looking at the XCM executor code, it looks like this fee mode only affects cases when additional fee must be paid. So XCM instructions that are affected by this setting are: `ExportMessage`, `LockAsset`, `RequestUnlock`, `TransferReserveAsset`, `DepositReserveAsset`, `InitiateReserveWithdraw`, `InitiateTeleport`, `QueryPallet`, `ReportError`, `ReportHolding`, `ReportTransactStatus`). And it looks like we don't execute any instructions from this set at `dest`.

The base execution fee is paid by the `BuyExecution` instruction, which ignores this setting. Or am i missing something?

Also e.g. `TransferReserveAsset` which is executed locally (at this chain) would prepend `xcm_on_dest` with `ClearOrigin` instruction and then if jit mode is on, it'll fail, because [it needs something in the origin register](https://github.com/paritytech/polkadot-sdk/blob/ff3a3bca449a3cae910d2dc8c7ed3658938c6478/polkadot/xcm/xcm-executor/src/lib.rs#L957).

Reply:
You're right, I was thinking to use `jit_withdraw: true` for `BuyExecution` as well, but I guess that'd be a more invasive change than just adding manual `WithdrawAsset(fees)` on dest chain where needed.
I am making dropping `SetFeesMode{ jit_withdraw: true }` in favor of `WithdrawAsset(fees)` now.

Reply:
hmm.. `WithdrawAsset` has same problem of `origin` being cleared + it'd have to use `beneficiary` account/origin anyway; looking into using `DescendOrigin` for this

Reply:
nope, `ClearOrigin` appended by the executor closes all avenues for getting the `fees` into holding on destination, I guess we're stuck with custom XCM alternative https://github.com/acatangiu/polkadot-sdk/commit/a80ff3c429313e25d6cadfbd0c17d8db022e42d0

Reply:
Yep, I also haven't found a workaround other than executing with overrided `SendXcm`, which would inject required instructions into XCM that is sent to destination. But your commit looks much better of course :)